dcos_installer/upgrade.py

```diff
@@ -0,0 +1,134 @@
+# limitations under the License.
+
+set -o errexit -o nounset -o pipefail
+
+source /opt/mesosphere/environment.export
+
+# Check if this is a terminal, and if colors are supported, set some basic
+# colors for outputs
+if [ -t 1 ]; then
+    colors_supported=$(tput colors)
+    if [[ $colors_supported -ge 8 ]]; then
+    RED='\e[1;31m'
+    BOLD='\e[1m'
+    NORMAL='\e[0m'
+    fi
+fi
+
+if [[ $EUID -ne 0 ]]; then
+    echo "This script must be run as root" 1>&2
+    exit 1
+fi
+
+
+# check for version of dc/os upgrading from
+version=`grep "version" /opt/mesosphere/etc/dcos-version.json | cut -d '"' -f 4`
+if [ $version != {{ installed_cluster_version }} ]; then
+   echo "ERROR: Expecting to upgrade DC/OS from {{ installed_cluster_version }} to {{ installer_version }}.\
+                Version found on node: $version"
+   exit 1
+fi
+
+echo "Upgrading DC/OS agent {{ installed_cluster_version }}  -> {{ installer_version }}"
+
+
+pkgpanda fetch --repository-url={{ bootstrap_url }} {{ cluster_packages }}
+
+
+pkgpanda activate --no-block {{ cluster_packages }}
+
+# check if we are on a master node
+if [ -f /etc/mesosphere/roles/master ]; then
+   # run exhibitor migration script here
+   until dcos-shell dcos-exhibitor-migrate-perform
+   do
+        status=$?
+        case $status in
+             1) echo "Waiting for exhibitor endpoint"
+                sleep 10
+                ;;
+             2) echo "Could not read from exhibitor"
+                echo "Contact Support"
+                exit
+                ;;
+             4) echo "Rolling update in progress"
+                exit
+                ;;
+             8) echo "At least one config value does not have the expected pre migration value,
+                      \ and automatic migration can not take place"
+                echo "Contact Support"
+                exit
+                ;;
+             16) echo "Attempting to start the rolling update failed due to a non 200 response from exhibitor"
+                 echo "Contact Support"
+                 exit
+                 ;;
+        esac
+   done
+fi
+
+"""
+
+
+def generate_node_upgrade_script(gen_out, installed_version, serve_dir=SERVE_DIR):
+
+    # installed_cluster_version: Current installed version on the cluster
+    # installer_version: Version we are upgrading to
+
+    bootstrap_url = gen_out.arguments['bootstrap_url']
+
+    installer_version = gen.calc.entry['must']['dcos_version']
+
+    package_list = ' '.join(package['id'] for package in gen_out.cluster_packages.values())
+
+    bash_script = gen.template.parse_str(node_upgrade_template).render({
+        'dcos_image_commit': util.dcos_image_commit,
+        'generation_date': util.template_generation_date,
+        'bootstrap_url': bootstrap_url,
+        'cluster_packages': package_list,
+        'installed_cluster_version': installed_version,
+        'installer_version': installer_version})
```

Comment:
Let's not have both `installed_version` and `installer_version` together. Those are really easy to mix up. I like `installer_version` as a name, it's as simple and direct as possible. This function uses both `installed_version` and `installed_cluster_version`. Let's pick one name for that and use it everywhere on this PR -- tests and function signatures and CLI help output included. My bias is toward `cluster_version` over `installed_cluster_version` since the latter is verbose, but whatever's clear and consistent and keeps the two versions distinguishable works for me.